libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,191 @@
+      future.cancel(false);
+      future = null;
+    }
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(Looper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        if (rawLocation != null) {
+          enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+            @Override
+            public void run() {
+              callback.onSuccess(LocationEngineResult.create(enhancedLocation));
+            }
+          });
+        }
+      }
+    }, 1500, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  void onNavigationKilled() {
+    if (future != null) {
+      removeLocationUpdates(callback);
+      future.cancel(false);
+      future = null;
+      executorService.shutdown();
+    }
+  }
+
+  void configure(@NonNull File path, @NonNull String version, @NonNull String host,
+                 @NonNull String accessToken) {
+    OfflineNavigator offlineNavigator =
+      new OfflineNavigator(mapboxNavigator.getNavigator(), version, host, accessToken);
+    String tilePath = new File(path, version).getAbsolutePath();
+    offlineNavigator.configure(tilePath, new OnOfflineTilesConfiguredCallback() {
+      @Override
+      public void onConfigured(int numberOfTiles) {
+        Timber.d("DEBUG: onConfigured %d", numberOfTiles);
+        onFreeDrive();
+      }
+
+      @Override
+      public void onConfigurationError(@NonNull OfflineError error) {
+        Timber.d("DEBUG: onConfigurationError %s", error.getMessage());
+      }
+    });
+  }
+
+  // TODO Remove when MapboxNavigation is refactored - added to get current tests to pass
+  LocationEngine getLocationEngine() {
+    return locationEngine;
+  }
+
+  @NonNull
+  private static LocationEngineRequest obtainLocationEngineRequest() {
+    return new LocationEngineRequest.Builder(UPDATE_INTERVAL_IN_MILLISECONDS)
+      .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+      .setFastestInterval(FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS)
+      .build();
+  }
+
+  private static Location getSnappedLocation(NavigationStatus status, Location fallbackLocation) {
+    Location snappedLocation = new Location(fallbackLocation);
+    snappedLocation.setProvider("enhanced");
+    FixLocation fixLocation = status.getLocation();
+    Point coordinate = fixLocation.getCoordinate();
+    snappedLocation.setLatitude(coordinate.latitude());
+    snappedLocation.setLongitude(coordinate.longitude());
+    if (fixLocation.getBearing() != null) {
+      snappedLocation.setBearing(fixLocation.getBearing());
+    }
+    snappedLocation.setTime(fixLocation.getTime().getTime());
+    return snappedLocation;
+  }
+
+  private void onLocationChanged(Location location) {
+    if (location != null) {
+      rawLocation = location;
+      mapboxNavigator.updateLocation(location);
```

Comment:
This is invoked on the main thread. Is it okay to update raw location on one thread and poll enhanced one on a different thread?

Reply:
not okay

Reply:
> not okay

Same approach is currently implemented in active guidance 👀 https://github.com/mapbox/mapbox-navigation-android/blob/master/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteProcessorBackgroundThread.kt and it's working 👌. Are we missing something?